### DISCUSSION TITLE
Getting the full error with post_p

### DISCUSSION THREAD

**Discussion by daleif** [mojolicious/mojo#2090] [General]
This morning my webapps had an issue. We're using `Mojolicious::Plugin::OAuth2` to auth users against Azure. Works perfectly fine, but not this time. Kept getting an `unauthorized` error. When I then messed with the code of `Mojolicious::Plugin::OAuth2` and changed a `post_p` to a normal post and looked at the response, the content field actually fully explained the error.

My question is this. When we're using a promise chain like

```
$c->oauth2->get_token_p($provider_name => \%get_token)
  ->then( sub {...  })
  ->catch(
		sub ($err) {
			    die 'error: ' . $err;
		}
);
```

do we then have any chance of accessing the full error in the `catch` branch? Or are we doomed to only being able to access whatever is given to the reject part of the promise.

Here `get_token_p` ends by returning a `post_p`.

**Comment by hussam-qasem** [comment]
How about:
1. Use `-async_await`. For example: `use Mojo::Base 'Mojolicious::Controller', -signatures, -async_await;`
2. Mark your controller as async. For example: `async sub my_sub ($self) {`
3. Use try/catch. Consider:
```perl
my $result = eval { await $c->oauth2->get_token_p($provider_name => \%get_token) };
if (my $err = $@) {
  die 'error: ' . $err;
}

# check $result and do_something()
```
